### DISCUSSION TITLE
binaries.prisma.sh Returns 404 Not Found Error

### DISCUSSION THREAD

**Discussion by jordan-bravo** [prisma/prisma#24405] [Q&A] [answered]
### Bug description

I'm using Prisma on NixOS, configured as per [this GitHub issue](https://github.com/prisma/prisma/issues/3026#issuecomment-2142940486), and things have worked properly in the past.  However, currently when the CLI tries to download the proper files for the prisma engine, it encounteres a "Not Found" error (see below).

```
❯ prisma --version
Warning Precompiled engine files are not available for nixos.
> Downloading Prisma engines for Node-API for linux-nixos [ ] 0% Error: Not Found https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/linux-nixos/migration-engine.gz
```

If I try to visit that [URL](https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/linux-nixos/migration-engine.gz) in the browser, it shows the 404 Not Found error.  In fact, the entire subdomain [binaries.prisma.sh](https://binaries.prisma.sh) is returning a 404 Not Found error.

This has been occurring for several days now so it doesn't seem to be a temporary outage.

This is preventing me from using prisma.   Any suggestions?   Any help is much appreciated.

### How to reproduce

1. Configure Nix Flake as per [this GitHub issue](https://github.com/prisma/prisma/issues/3026#issuecomment-2142940486).
2. Run any prisma command such as `prisma --version`.

### Expected behavior

The appropriate prisma binaries should be downloaded from binaries.prisma.sh and prisma should run.

### Prisma information

N/A

### Environment & setup

- OS: NixOS
- Database: N/A (but using Postgres)
- Node.js version: N/A (using v18 but same issue occurs with v20)

### Prisma Version

Can't run prisma commands, see above error.

**Comment by janpio** [comment]
`binaries.prisma.sh` is working as intended, but we do not build our engines for Nix and NixOS - so there are not engine files hosted there.

Usually following the instructions from that issue allows you to work around that. Seems something goes wrong when you do.

**Reply by jordan-bravo** [reply]
I'm confused.  Why is it trying to download a binary from that URL?
```
❯ prisma --version
Warning Precompiled engine files are not available for nixos.
> Downloading Prisma engines for Node-API for linux-nixos [                    ] 0%Error: Not Found https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/linux-nixos/migration-engine.gz
```

Any suggestions on what might be causing the problem?

Edit:  Are you sure that the issue isn't related to prisma's binaries server returning a 404 error?  

I tried a different flake setup copied directly from the README in [nix-prisma-utils](https://github.com/VanCoding/nix-prisma-utils) and the error is similar:

```
error: builder for '/nix/store/iyr3n3hr1ii17nwh518cr61mpxncn3h6-https-binaries.prisma.sh-all_commits-4bc8b6e1b66cb932731fb1bdbbc550d1e010de81-debian-openssl-3.0.x-schema-engine.gz.drv' failed with exit code 1;
       last 7 log lines:
       >
       > trying https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/debian-openssl-3.0.x/schema-engine.gz
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (56) The requested URL returned error: 404
       > error: cannot download https-binaries.prisma.sh-all_commits-4bc8b6e1b66cb932731fb1bdbbc550d1e010de81-debian-openssl-3.0.x-schema-engine.gz from any mirror
```

It seems like there's a problem with the server because it returns a 404 on several URLs.

**Reply by janpio** [reply]
Yes, URLs that are supposed to exist work fine, for example: https://binaries.prisma.sh/all_commits/e9771e62de70f79a5e1c604a2d7c8e2a0a874b48/debian-openssl-3.0.x/schema-engine.gz (schema engine for Debian OpenSSL 3.0.x of 5.14.0)

I unfortunately don't know about Nix and NixOS to answer your questions. I'll see if I can find someone who knows more.

**Comment by aqrln** [comment]
Do you actually use the flake (i.e. do you enter it via `nix develop` or with `use flake` in direnv)?

**Reply by jordan-bravo** [reply]
Yes, I use nix-direnv with `use-flake`.  I have also tried disabling direnv and using the command `nix develop`, but the result is the same.

**Reply by aqrln** [reply]
Can you post your `flake.nix` please?

**Reply by aqrln** [reply]
and the output of `env | grep PRISMA` from inside your nix shell

**Reply by jordan-bravo** [reply]
Hi, apologies for the late reply.  Thank you for taking a look at this.

Yes, I will provide the requested info below.  However, I have an update:

I ran the Nix garbage collector and rebooted.  Now, I can successfully run some prisma commands, such as `prisma --version`, however the generate command is still throwing an error, for example `prisma generate` outputs the following:

```
❯ prisma generate
Prisma schema loaded from prisma/schema.prisma
[Error: ENOENT: no such file or directory, mkdir '/nix/store/vqfdiw702jsrnnhdskwswm805jyigz6w-prisma-5.12.1/lib/node_modules/prisma/engines'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'mkdir',
  path: '/nix/store/vqfdiw702jsrnnhdskwswm805jyigz6w-prisma-5.12.1/lib/node_modules/prisma/engines'
}
```

Here is my `flake.nix`:
```nix
{
  description = "This Nix flake creates a development shell for the Customer API that provides the
  required versions of dependencies such as NodeJS 18";
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
  inputs.flake-utils.url = "github:numtide/flake-utils";

  outputs = { nixpkgs, flake-utils, ... }:
    flake-utils.lib.eachDefaultSystem (system: let
      pkgs = nixpkgs.legacyPackages.${system};
    in {
      devShells.default = pkgs.mkShell {
        buildInputs = with pkgs; [
          nodejs_18
          infisical
          nodePackages.prisma
          openssl
          prisma-engines
        ];
        shellHook = with pkgs; ''
          # Prisma related
          export PRISMA_SCHEMA_ENGINE_BINARY="${prisma-engines}/bin/schema-engine"
          export PRISMA_QUERY_ENGINE_BINARY="${prisma-engines}/bin/query-engine"
          export PRISMA_QUERY_ENGINE_LIBRARY="${prisma-engines}/lib/libquery_engine.node"
          export PRISMA_INTROSPECTION_ENGINE_BINARY="${prisma-engines}/bin/introspection-engine"
          export PRISMA_FMT_BINARY="${prisma-engines}/bin/prisma-fmt"
          export PATH="$PWD/node_modules/.bin/:$PATH"
        '';
      };
    });
}
```

And here is the output of `env | grep PRISMA`:
```shell
export PRISMA_SCHEMA_ENGINE_BINARY="/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/bin/schema-engine"
export PRISMA_QUERY_ENGINE_BINARY="/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/bin/query-engine"
export PRISMA_QUERY_ENGINE_LIBRARY="/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/lib/libquery_engine.node"
export PRISMA_INTROSPECTION_ENGINE_BINARY="/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/bin/introspection-engine"
export PRISMA_FMT_BINARY="/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/bin/prisma-fmt"
PRISMA_SCHEMA_ENGINE_BINARY=/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/bin/schema-engine
PRISMA_INTROSPECTION_ENGINE_BINARY=/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/bin/introspection-engine
PRISMA_FMT_BINARY=/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/bin/prisma-fmt
PRISMA_QUERY_ENGINE_BINARY=/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/bin/query-engine
PRISMA_QUERY_ENGINE_LIBRARY=/nix/store/nbpdijvx4lkf89d1vwiq3wihibxj3ddx-prisma-engines-5.12.1/lib/libquery_engine.node
```

**Reply by jordan-bravo** [reply]
Hi, just wanted to ask if there are any suggestions with this?  Thank you.

**Reply by aqrln** [reply]
Does `npx prisma generate` work if you only use `prisma-engines` from nixpkgs but not `nodePackages.prisma` and get the normal Prisma CLI from npm?

**Reply by jordan-bravo** [reply]
I removed `nodePackages.prisma`, reloaded the environment, and ran `npm install`.  Then,

```
❯ npx prisma generate
Warning Precompiled engine files are not available for nixos.
> Downloading Prisma engines for Node-API for linux-nixos [ ] 0%
Error: Failed to fetch sha256 checksum at https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/linux-nixos/migration-engine.sha256. 404 Not Found

If you need to ignore this error (e.g. in an offline environment), set the PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING environment variable to a truthy value.
Example: PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING=1
```

If I follow the suggestion, it still throws a similar error:
```
❯ PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING=1 npx prisma generate
Warning Precompiled engine files are not available for nixos.
> Downloading Prisma engines for Node-API for linux-nixos [ ] 0%
Error: Not Found https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/linux-nixos/migration-engine.gz
````

**Reply by aqrln** [reply]
Oh, I see:
> https://binaries.prisma.sh/all_commits/4bc8b6e1b66cb932731fb1bdbbc550d1e010de81/linux-nixos/migration-engine.sha256

You are on Prisma 4.x (4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 corresponds to 4.16 if I'm reading the git tags correctly) and it fails to download `migration-engine` that doesn't exist anymore in modern Prisma (as it was renamed to `schema-engine`) and thus isn't provided by the `prisma-engines` Nix package and the environment variables you have set.

If you want to use `prisma-engines` from nixpkgs, it needs to correspond exactly to the version of Prisma you have in your project. If you rely on a legacy version of Prisma and can't update for any reason yet, you can pin nixpkgs to an old commit that has the version you need. Or you can use https://github.com/VanCoding/nix-prisma-utils to use any version of Prisma on NixOS, whether older or newer than in nixpkgs.

**Reply by jordan-bravo** [reply, marked as answer]
Got it, that was the issue.  I pinned nixpkgs to a revision with prisma v4.x.y and now it works properly.  Here is my working flake, in case it's helpful to anyone in the future:
```nix
{
  description = "This Nix flake creates a development shell for the Customer API that provides the
  required versions of dependencies such as NodeJS 18 and Prisma 4";
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
    nixpkgs-prisma-4.url = "github:NixOS/nixpkgs/29bcead8405cfe4c00085843eb372cc43837bb9d";
    flake-utils.url = "github:numtide/flake-utils";
  };

  outputs = { nixpkgs, nixpkgs-prisma-4, flake-utils, ... }:
    flake-utils.lib.eachDefaultSystem (system: let
      pkgs = nixpkgs.legacyPackages.${system};
      pkgs-prisma-4 = nixpkgs-prisma-4.legacyPackages.${system};
    in {
      devShells.default = pkgs.mkShell {
        nativeBuildInputs = [ pkgs.bashInteractive ];
        buildInputs = [
          pkgs.nodejs_18
          pkgs-prisma-4.nodePackages.prisma
          pkgs-prisma-4.prisma-engines
        ];
        shellHook = with pkgs-prisma-4; ''
          export PRISMA_MIGRATION_ENGINE_BINARY="${prisma-engines}/bin/migration-engine"
          export PRISMA_SCHEMA_ENGINE_BINARY="${prisma-engines}/bin/schema-engine"
          export PRISMA_QUERY_ENGINE_BINARY="${prisma-engines}/bin/query-engine"
          export PRISMA_QUERY_ENGINE_LIBRARY="${prisma-engines}/lib/libquery_engine.node"
          export PRISMA_INTROSPECTION_ENGINE_BINARY="${prisma-engines}/bin/introspection-engine"
          export PRISMA_FMT_BINARY="${prisma-engines}/bin/prisma-fmt"
          export PATH="$PWD/node_modules/.bin:$PATH"
        '';
      };
    });
}
```
